src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -235,15 +236,42 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
       vds
   }
 
-  def exportGen(path: String, precision: Int = 4) {
-    requireSplit("export gen")
+  def exportBgen(path: String, nBitsPerProb: Int = 8) {
+    requireSplit("export bgen")
 
-    def writeSampleFile() {
-      //FIXME: should output all relevant sample annotations such as phenotype, gender, ...
-      vds.hc.hadoopConf.writeTable(path + ".sample",
-        "ID_1 ID_2 missing" :: "0 0 0" :: vds.sampleIds.map(s => s"$s $s 0").toList)
-    }
+    val nSamples = vds.nSamples
+
+    val rsidQuery: Option[Querier] = vds.vaSignature.getOption("rsid")
+      .filter {
+        case TString => true
+        case t => warn(
+          s"""found `rsid' field, but it was an unexpected type `$t'.  Emitting missing rsID.
+             |  Expected ${ TString }""".stripMargin)
+          false
+      }.map(_ => vds.queryVA("va.rsid")._2)
+
+    val varidQuery: Option[Querier] = vds.vaSignature.getOption("varid")
+      .filter {
+        case TString => true
+        case t => warn(
+          s"""found `varid' field, but it was an unexpected type `$t'.  Emitting missing variant ID.
+             |  Expected ${ TString }""".stripMargin)
+          false
+      }.map(_ => vds.queryVA("va.varid")._2)
+
+    val bgenHeader = BgenWriter.emitHeaderBlock(vds)
+
+    vds.rdd.mapPartitions { it: Iterator[(Variant, (Annotation, Iterable[Genotype]))] =>
+      it.map { case (v, (va, gs)) =>
+        BgenWriter.emitVariantBlock(v, va, gs, rsidQuery, varidQuery, nSamples, nBitsPerProb)
+      }
+    }.saveFromByteArrays(path + ".bgen", vds.hc.tmpDir, header = Some(bgenHeader))
```

Comment:
OK, we need to modify this to support a `export_bgen(..., parallel=True)` option that writes one BGEN file per partition, since the call to `copyMerge` in `saveFromByeArrays` will kill us in creating multiple copies of UKBBv1.

Second, we should create a single ArrayBuilder inside the mapPartitions here and pass it into emitVariantBlock so you don't create an ArrayBuilder for each variant.